controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -93,6 +93,47 @@ public void processingFailure(Exception error) {
         return result;
     }
 
+    public CompletableFuture<Boolean> truncateSegment(final String scope,
+                                                    final String stream,
+                                                    final int segmentNumber,
+                                                    final long offset,
+                                                    final HostControllerStore hostControllerStore,
+                                                    final ConnectionFactory clientCF) {
+        final CompletableFuture<Boolean> result = new CompletableFuture<>();
+        final Controller.NodeUri uri = getSegmentUri(scope, stream, segmentNumber, hostControllerStore);
+
+        final WireCommandType type = WireCommandType.TRUNCATE_SEGMENT;
+
+        final FailingReplyProcessor replyProcessor = new FailingReplyProcessor() {
+
+            @Override
+            public void connectionDropped() {
+                result.completeExceptionally(
+                        new WireCommandFailedException(type, WireCommandFailedException.Reason.ConnectionDropped));
+            }
+
+            @Override
+            public void wrongHost(WireCommands.WrongHost wrongHost) {
+                result.completeExceptionally(new WireCommandFailedException(type, WireCommandFailedException.Reason.UnknownHost));
+            }
+
+            @Override
+            public void segmentTruncated(WireCommands.SegmentTruncated segmentTruncated) {
+                result.complete(true);
+            }
+
+            @Override
+            public void processingFailure(Exception error) {
+                result.completeExceptionally(error);
+            }
+        };
+
+        WireCommands.TruncateSegment request = new WireCommands.TruncateSegment(idGenerator.get(),
```

Comment:
What's the expected behavior if we ask the segment store to truncate it twice?

Reply:
A truncated segment can be truncated multiple times. 
Expectation is that as long as we are truncating it at a point gt eq previous truncation point, the truncation should succeed. 

We want to do `gt eq` and not strictly gt because imagine a scenarion where we have a segment that is never written to and never seals either. In each new stream cut we will always have this segment with same offset (0). 
We should be able to tell segmentStore to trim this segment multiple times at the same offset.

Reply:
Hmm, I'm not sure about the implementation of replyProcessor right above this. It expects the segment store to respond `SegmentTrucated` to complete positively. However, looking at the implementation in `PravegaRequestProcessor`, I'm not sure it returns `SegmentTruncated` in the case we ask it to truncate twice the same offset. Could you double check, please?

Reply:
StreamSegmentStore.java

    /**
     * Truncates a StreamSegment at a given offset.
     *
     * @param streamSegmentName The name of the StreamSegment to truncate.
     * @param offset            The offset at which to truncate. This must be at least equal to the existing truncation
     *                          offset and no larger than the StreamSegment's length. After the operation is complete,
     *                          no offsets below this one will be accessible anymore.
     * @param timeout           Timeout for the operation.
     * @return A CompletableFuture that, when completed normally, will contain the new length of the StreamSegment.
     * If the operation failed, the future will be failed with the causing exception.
     */
    CompletableFuture<Long> truncateStreamSegment(String streamSegmentName, long offset, Duration timeout);

Reply:
it says at least equal to existing truncation offset

Reply:
The question is about requesting to truncate the same offset multiple times. The first time will work fine. For the second time, the truncation segment will be larger and I believe this is going to return an error.

Reply:
The length of the Segment has nothing to do with truncation offset. @shiveshr is correct. If we truncate a Segment at the same offset it was previously truncated, the operation will still succeed (so it's idempotent from that POV). If the new truncation offset is less than the previous one, then an exception will be thrown (`BadOffsetException`).

Check out the actual code here: https://github.com/pravega/pravega/blob/master/segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/SegmentMetadataUpdateTransaction.java#L320

Reply:
Ok, thanks for the feedback.